lib/astarte_core/triggers/policy/range_error.ex

```diff
@@ -0,0 +1,49 @@
+#
+# This file is part of Astarte.
+#
+# Copyright 2021 Ispirata Srl
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+defmodule Astarte.Core.Triggers.Policy.RangeError do
+  use Ecto.Schema
+  import Ecto.Changeset
+
+  alias Astarte.Core.Triggers.Policy.RangeError
+
+  @error_range 400..599
+
+  embedded_schema do
+    field :error_codes, {:array, :integer}
+  end
+
+  def validate(changeset) do
+    changeset
+    |> validate_subset(:error_codes, @error_range)
+    |> validate_length(:error_codes, min: 1)
+  end
+
+  defimpl Jason.Encoder, for: RangeError do
+    def encode(%RangeError{} = range_error, options) do
+      %RangeError{
+        error_codes: error_codes
+      } = range_error
+
+      %{
+        error_codes: error_codes
+      }
+      |> Jason.Encoder.Map.encode(options)
+    end
+  end
```

Comment:
Since the encode functions is doing plain struct -> map translation you can use `@derive` instead, see the second and third blocks [here](https://github.com/michalmuskala/jason/blob/master/README.md#differences-to-poison)

In your case the `@derive` should be placed before `embedded_schema` since that's where `defstruct` is happening (via a macro).